SalesforceCommerceCloud/pwa-kit
# [V2] Address stale browser caching between projects/templates

# Description

Ticket W-11087303

This PR attempts to minimize the issue with browser caching localhost (during development). Recently, the homepage for Retail React App has a **cache-control header** of 1 year. When you run `npm start` on a _different_ template/project, you'd notice that the browser would use its caching of the homepage from previous template, because after all, it's still the same url of `localhost:3000`. In other words, during development, it's possible that this cache header "leaks" onto other templates.

Initially, we suspect that the root cause might be the **service worker** (SW). But...
- with SW removed and homepage's cache header set: the issue was still there
- with SW installed and homepage's cache header not set: the issue was gone

So I concluded that it's the cache header (and browser caching) that is the root cause.

## Strategy to minimize the issue

Given these criteria:
- minimal, non-intrusive: developers still have control
- visible: developers notice it and know what’s causing it

...this PR proposes doing 2 things to minimize the caching issue:
- Setting a **shorter caching time** for the Retail React App's homepage ← 15 mins is a good start IMO.. users can choose to replace it with something longer for their own business use case 
- ~~Setting the **[Clear-Site-Data header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data)** to tell browsers to clear their cache. And it's set on the loading page, before landing on the template's homepage. The [browser support](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Clear-Site-Data#browser_compatibility) for this header is good enough, especially for development purpose. It'll be like a nice enhancement for those browsers that do support it. For clearing _cache_, FYI Firefox and Safari do not support it.~~
- The loading page would redirect to homepage's url that's appended with the **project name as cache breaker**

## Alternatives

Ben has mentioned here in an [old PR](https://github.com/SalesforceCommerceCloud/pwa-kit/pull/103#discussion_r709420555) about **overriding/ignoring cache header** during development. I think it's still nice to be able to see this header in action when in development.

Another option would be: developers can **turn on the "disable cache" option** in Chrome's inspector on their end. This is typically suggested online by other folks.

Also, another option is we open the project under **Chrome's incognito window**, which is possible given the `open` library we use. However, `open` supports only Chrome version of incognito window and not other browsers, which IMO very intrusive for those developers who prefer another browser like Firefox, for example. Also, incognito window has some limitations (e.g. [storage limitation](https://developer.chrome.com/docs/workbox/understanding-storage-quota/#special-chrome-incognito-considerations)) that developers may not like. In other words, I think the decision of whether to use incognito window should be left up to the developers.

# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# How to Test-Drive This PR

0. Clear your browser cache
1. Run `npm ci` at the monorepo root
2. `npm start` in the `template-retail-react-app` folder
3. Wait for the homepage to appear
4. Kill your local dev server
5. `npm start` in the `template-typescript-minimal` folder
6. Wait for the homepage to appear
7. Verify that this time it's a different homepage that is loaded

# Checklists


## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
